### PR TITLE
Added `DataView.getAllSelectedRows` to allow getting all the selected rows on all the pages

### PR DESCRIPTION
**Pull request by arashdalir**
**scenario:**
there are 1000 entries on your table, you have columns like `name`, `e-mail`, `location`, etc. you want to select all who are in `Vienna`, `Berlin` and `Rome` and get their emails. the number of rows with this criterium is around 50 and they are distributed over multiple pages; each page containing 50. you filter the location to `Vienna`, select all Viennese, then you filter `Berlin` and select them, and then do the same with `Rome`. now you delete your filter and only 1 person is selected on the active page you are seeing.
if you use `grid.getSelectedRows()` or `grid.getSelectionModel().getSelectedRows()` you will only get the `idx` of that 1 selected person of the active page. but wait a minute! you selected 50; how can you get all without needing to filter again? and what if you don't want to filter the rows, because you don't want to change your current view (filtered something else, or whatever!)?
well! damn! you can't! the information which rows are actually selected, is inside the `Slick.Data.DataView.syncGridSelection` function, and not even available to the `DataView` itself! so there is no way to tell the user that they have selected 50 over all the pages and only 1 on the active page (well, the second part is possible, the first one is not!)

**solution**:
get the `selectedRowIds` out of the `syncGridSelection` and into `DataView`, then create the `getSelectedRows` function and export it for general use!

### CODE REVIEW

File: composer.json
Comment (ghiscoding) [inline composer.json:1]: please remove this file, it shouldn't be included in this PR

File: slick.dataview.js
Comment (ghiscoding) [inline slick.dataview.js:1056]: you should move this variable declaration to the top (around line 74)

**Review by ghiscoding — CHANGES_REQUESTED** [review]

File: slick.dataview.js
Comment (arashdalir) [reply slick.dataview.js:1056]: sure, will do!

File: composer.json
Comment (arashdalir) [reply composer.json:1]: oops! it shouldn't have been committed, sorry...

**Comment by ghiscoding** [comment]
I actually wanted that feature, so I'll test it out in coming days, too bad you missed the window for a new release we did one yesterday.

**Comment by arashdalir** [comment]
sorry mate! I just realized this issue exists today! :D

tbh, I believe this is not the right place to do this; because there are 2 other `getSelectedRows` functions, one in `SlickGrid` and one in `RowSelectionModel`; so this makes the code a bit more complicated and harder to read and follow. But as this is the only place the complete list of selected rows is stored, I found no other "fast" solutions for it.

the other, "better" solution would be to rewrite the `syncGridSelection` function and move into either one of those mentioned classes and make them responsible for the synchronization; but well, that's a bigger change which is also incompatible with current versions and deployments, meaning it will break everyone's code...

**Comment by ghiscoding** [comment]
Right, thanks for pointing that out, I forgot that `getSelectedRows()` is from SlickGrid not DataView, the other thing also is that not everyone uses the DataView so we have to take that in consideration as well. However, to cover the fact that not everyone uses the DataView, most of the DataView methods names are slightly different compare to SlickGrid (core) so that no methods are in conflict, so perhaps you can change the name slightly, not sure what name to use though. 

As for Rewriting the SlickGrid `getSelectedRows` method in the core, that actually might not (probably won't) work because when you don't use the DataView that means you use only the SlickGrid core only and you have the entire dataset in-memory and I don't think Pagination works without the DataView... so having this new method in the DataView might still be the best place to do so but again I think we should use a different name to avoid conflicts and tons of questions of why it doesn't work

**Comment by arashdalir** [comment]
> most of the DataView methods names are slightly different compare to SlickGrid (core) so that no methods are in conflict, so perhaps you can change the name slightly, not sure what name to use though.

I'd say as the function is not bound to SlickGrid core and must be accessed from `DataView` instances, the naming wouldn't conflict and it makes - imho - sense, because then to get both variations, one can use `grid.getSelectedRows()` and `dataView.getSelectedRows()`. But if you think this should be done, I'll change the name to something like `getAllSelectedRows`...

**Comment by ghiscoding** [comment]
Yes I would prefer a different name, I meant to say "confusions" not "conflicts"... I'm almost sure that is why the author didn't use the same method names and I do like the new name `getAllSelectedRows`

**Comment by 6pac** [comment]
this looks good, it's ready to merge?

**Review by ghiscoding — APPROVED** [review]

**Comment by ghiscoding** [comment]
@6pac 
I haven't had a chance to try it yet but the code looks good after latest changes. 
You can merge as I guess it's ok, will try in coming days

**Comment by arashdalir** [comment]
well, I cannot merge as I don't have any write access! :D

but no worries! not in a hurry! using the main branch in my developments, so I don't have any issues locally. 

image

**Comment by ghiscoding** [comment]
@arashdalir @6pac 
So I was testing this today and what I mentioned earlier (possible confusion), well it happened to me lol... but since I'm using TypeScipt and I added `getAllSelectedRows` method only to the DataView then TS caught it right away and told me that "getAllSelectedRows" doesn't exist on SlickGrid".. so all that to say that it's really great to keep different method naming between Slick.Grid and Slick.DataView.

However, I'm contemplating adding another method in your PR that would return not just the Selected Ids but also the Item DataContext objects, most people will do it anyway so why not add it directly the lib. I see 2 ways of doing it
1. modify the `getAllSelectedRows` that could take another argument and if it's enabled then return appropriately. 
```js
function getAllSelectedRows(isReturningDataContext) {
   if (isReturningDataContext) {
     return selectedDataContexts;
  } else {
    return selectedRowIds;
  }
}
```
2. or simply add another method to return all selected dataContext
```js
function getAllSelectedDataContexts() {
  return selectedDataContexts;
}
```
I'm more in favor of option 2 to add another method since option 1 would get cluttered and would return a different output (an array of indexes or an array of objects)

EDIT

oops @6pac just merged it lol... will do another PR then but anyway I'd like to have both your inputs

**Comment by arashdalir** [comment]
hi @6pac! thanks for the review!

I'd say I am in favor of the second solution too, because this function will then have the same prototype as the `getSelectedRows` functions. I actually needed to get the records out of `dataView`, but to ensure compatibility with the `getSelectedRows`, I used something like the following code. I suppose `selectedDataContexts` will return the same thing, but well, then I'd say if we add this function to `DataView`, it would be nice to add a similar one to `SlickGrid` and `RowSelectionModel`, wouldn't it?

```Javascript
for(var r = 0; r < rows.length; r++)
{
      var item = dataView.getItemByIdx(rows[r]);
//... do the rest
}
```

**Comment by 6pac** [comment]
whoops. yep, option 2.

**Comment by ghiscoding** [comment]
@arashdalir 
I actually can't get it to work with Pagination, I must be missing something. I have `syncGridSelection` enabled is there something else that I need for this to work?

**Comment by arashdalir** [comment]
could you please post your grid and dataview configs? maybe you are using `cellSelectionModel`?

the following code is the part which saves the list when `preserveHiddenOnSelectionChange` is set to true and grid is set to work in `multiselect` mode!

```javascript
grid.onSelectedRowsChanged.subscribe(function (e, args) {
        if (inHandler) { return; }
        var newSelectedRowIds = self.mapRowsToIds(grid.getSelectedRows());
        if (!preserveHiddenOnSelectionChange || !grid.getOptions().multiSelect) {
          setSelectedRowIds(newSelectedRowIds);
        } else {
          // keep the ones that are hidden
          var existing = $.grep(selectedRowIds, function (id) { return self.getRowById(id) === undefined; });
          // add the newly selected ones
          setSelectedRowIds(existing.concat(newSelectedRowIds));
        }
      });
```

**Comment by ghiscoding** [comment]
ah ok, I'm not sure why but I had `multiSelect` set to False for some reason and now it works when set to True, this is great.

So I got the new method working, you can maybe test it out and provide feedback
Also note that to test it, I provided Ids that are different to row indexes (by the way, `getAllSelectedRows` returns Ids not Rows indices like SlickGrid does, so we might be better renaming it to `getAllSelectedIds` as that confused me at first when testing my new method)

```js
function getAllSelectedItems() {
  var selectedData = [];
  var selectedIds = getAllSelectedRows();
  selectedIds.forEach(function (id) {
      selectedData.push(self.getItemById(id));
  });
  return selectedData;
}
```

**Comment by ghiscoding** [comment]
@arashdalir @6pac 
Another Possible Second Improvement

The previous [comment](https://github.com/6pac/SlickGrid/pull/542#issuecomment-709411278) also makes me think it might be nice to review `slick.checkboxselectcolumn.js` and possibly make it to optionally use Slick.DataView because it behaves differently with Pagination, for example on my side the users get confused when they click "Select All" they image that it's for the entire dataset but instead it's only applied for current page because that plugin uses `grid.getSelectedRows()` and `isSelectAllChecked = selectedRows.length >= grid.getDataLength()` to know if everything is checked (which again is only valid for current page).

1. We could maybe do something like this
```js
var isUsingDataView = !Array.isArray(this.getData()); // getData returns grid data array or the DataView object
var selectedIds = dataView.getSelectedIds();
var isSelectAllChecked = selectedIds.length >= dataView.getLength();
```
2. or if we want to know if it's all selected in current page **and** if it's all selected in the entire dataset then keep 2 flags
```js
var isUsingDataView = !Array.isArray(this.getData());
var selectedIds = dataView.getSelectedIds();
var selectedRows = grid.getSelectedRows(); // current page rows

var isSelectAllCurrentPageChecked = selectedIds.length >= grid.getDataLength();
var isSelectAllDatasetChecked = selectedIds.length >= dataView.getLength();
```
If going with number 2, we could maybe use a tri-state icon [-] when `isSelectAllCurrentPageChecked` is true but not `isSelectAllDatasetChecked` and if both are true then [x]

#### selected all current page
image
#### selected all entire dataset
image

I think I'm going to push for the number 2 further and eventually do a PR (actually 2 PRs because the previous [comment](https://github.com/6pac/SlickGrid/pull/542#issuecomment-709411278) highlights the first small change to add `getAllSelectedItems()`). However this could break some people's code, I mean if yesterday it was selecting only current page and tomorrow we change it so that it selects the entire dataset on all pages, we could maybe add a flag to go for behavior 1 or 2,

**Comment by arashdalir** [comment]
@ghiscoding 
> Also note that to test it properly, I made sure to have Ids not equal to row indexes (by the way, getAllSelectedRows returns Ids not Rows indices like SlickGrid does, so we might be better renaming your function to getAllSelectedIds since that confused me at first when implementing my new function).

well, AFAIK and have tried it (it is practically working), it still returns the indexes, i.e. row offset - and not the IDs (please refer to my working example); so I believe when, then it should be named getAllSelectedIdx

@6pac 
> The previous comment also makes me think it might be nice to review slick.checkboxselectcolumn.js and possibly make it to optionally use Slick.DataView because it behaves differently with Pagination, for example on my side the users get confused when they click "Select All" they image that it's for the entire dataset but instead it's only applied for current page because that plugin uses grid.getSelectedRows() and isSelectAllChecked = selectedRows.length >= grid.getDataLength() to know if everything is checked (which again is only valid for current page).

well, the thing is, this behavior is something which user should be able to decide! I have a confirmation box on my implementation for cases where the number of selectedRows are more the number of shown selected rows; that is, if `grid.getSelectedRows()` is a subset of `dataView.getAllSelectedRows()`, and the user should decide if they want to have the action done to visible selected or all selected. Maybe I missed your point, but I would say, it's something the end-user should decide.  
The point with [-] is a good suggestion! I would say it will come in handy. But we need to be careful with the 3-state click thing; because it would get really complicated really soon! you would expect to be able to unselect a page or subset - after filtering - and the way you described the functionality, it will never be possible to select based on filter 1, then use filter 2 to unselect a few. I would really say we should create the proper API calls for it, but let the developers implement their own solution for "(un)select all on all pages"...

**Comment by ghiscoding** [comment]
@arashdalir 
> well, AFAIK and have tried it (it is practically working), it still returns the indexes, i.e. row offset - and not the IDs (please refer to my working example); so I believe when, then it should be named getAllSelectedIdx

No it doesn't, it really is the selected ids and the variable name gives it away, you are calling `return selectedRowIds;`, they are IDs not row indexes.

**Comment by 6pac** [comment]
I'm struggling a bit to see the use case for selecting all the rows on a page.  It seems very unlikely that precisely one page of records would need selection.  
Surely a user would expect the 'select all' checkbox to select everything?

I can see a logic in doing a 'select all filtered rows' where a row filter was applied.

**Comment by ghiscoding** [comment]
That's the point, the "Select All" doesn't select all when using Pagination, it only select all row of current page and that is what confuses most people. 

See this animated gif for a better explanation, most interestingly is the "Select All" checkbox on top left

VjkDONJsqW

**Comment by 6pac** [comment]
So we'd be giving the option for that behaviour just to make the change non-breaking? I think I'd be happy just to change the behaviour to select everything, even if it was breaking.

**Comment by ghiscoding** [comment]
I would rather do that with flags, this would change the behavior quite a lot. I know selecting everything is expected but some users might have bunch of grids with that behavior so...

**Comment by 6pac** [comment]
OK then, I see your point :-)

**Comment by arashdalir** [comment]
@ghiscoding oh I see my problem now! in my dataset the idx was the same as id; confused me...

please go for the function name change :) you are right, again!

**Comment by arashdalir** [comment]
@6pac 
> I'm struggling a bit to see the use case for selecting all the rows on a page. It seems very unlikely that precisely one page of records would need selection.

imagine writing an email-list. imagine using gmail. would you like to select all your thousends of e-mails when seeing those on page 1 of the list and then (maybe unintentionally) delete all of them? this might happen if a user clicks on this feature without being notified of the consequences...

**Comment by ghiscoding** [comment]
which is why I was thinking about a possible (optional) tri-state click that I wrote in this [post](https://github.com/6pac/SlickGrid/pull/542#issuecomment-709557314) with something like this

#### selected all current page
image
#### selected all entire dataset
image

Not every users have the same expectation, the email is a good example however that is not local pagination (it's backend pagination). I would rather have this tri-state click working with local pagination (in-memory grid/data).

**Pull request merged**
